### PR TITLE
Check for multiple module declarations in file

### LINKED ISSUE #130: Linter does not report more than one module per file
As per first bullet in [Basic Style Elements](https://github.com/lowRISC/style-guides/blob/master/VerilogCodingStyle.md#basic-style-elements) section of the style guide:
* Use SystemVerilog-2012 conventions, files named as module.sv, **one file per module**

There is no corresponding linter rule.

Example:
```verilog
module first;
endmodule
module second;
endmodule
```

### PR DESCRIPTION
**Pull request by rkapuscik**
Fixes #130.

### CODE REVIEW

File: verilog/tools/lint/testdata/defparam_usage.sv [hunk 1]
```diff
@@ -1,3 +1,4 @@
+/// verilog_lint: waive-start one-module-per-file
 module defparam_usage #(parameter int MY_PARAM = 0);
```
Comment (fangism) [inline verilog/tools/lint/testdata/defparam_usage.sv]: Instead of waiving, we could just delete the first definition.  This module definition is not actually needed to trigger the rule.

File: verilog/analysis/checkers/one_module_per_file_rule_test.cc [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+       "module ", {kToken, "moo"}, "; endmodule"},
+      {"class c; endclass\n"
+       "module foo; endmodule\n"
+       "module ", {kToken, "moo"}, "; endmodule"},
+      {"package q; endpackage\n"
+       "module foo; endmodule\n"
+       "module ", {kToken, "moo"}, "; endmodule\n"},
+  };
+  RunLintTestCases<VerilogAnalyzer, OneModulePerFileRule>(kTestCases, "");
+}
```
Comment (fangism) [inline verilog/analysis/checkers/one_module_per_file_rule_test.cc]: According to the LRM, it looks like module_declaration is allowed as a 'non_port_module_item', which suggests that modules may be nested.  I think we may want to restrict the findings to only outermost module declarations.
Suggested test case:

```systemverilog
module outer;
module inner;
endmodule : inner
endmodule : outer
```

The `module-filename` rule probably needs to be updated to account for this as well, but that can be a separate PR if you like.

File: verilog/tools/lint/testdata/defparam_usage.sv (hunk 1 above)
Comment (rkapuscik) [reply verilog/tools/lint/testdata/defparam_usage.sv]: Changed.

File: verilog/analysis/checkers/one_module_per_file_rule_test.cc (hunk 2 above)
Comment (rkapuscik) [reply verilog/analysis/checkers/one_module_per_file_rule_test.cc]: Good find, thanks! I've adapted both rules to account for that.

File: verilog/analysis/checkers/module_filename_rule.cc [hunk 3]
```diff
@@ -73,6 +74,14 @@ void ModuleFilenameRule::Lint(const TextStructureView& text_structure,
...
   auto module_matches = FindAllModuleDeclarations(*tree);
 
+  // Remove nested module declarations
+  module_matches.erase(std::remove_if(module_matches.begin(),
+              module_matches.end(),
+              [](verible::TreeSearchMatch& m) {
+                  return m.context.IsInside(NodeEnum::kModuleDeclaration);
+              }),
+          module_matches.end());
+
```
Comment (fangism) [inline verilog/analysis/checkers/module_filename_rule.cc]: Avoid vector::erase() in this case for its potential O(N^2) cost.
Rewrite as `std::copy_if` or `std::remove_copy_if` using a temporary vector.

File: verilog/analysis/checkers/one_module_per_file_rule.cc [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+  auto module_matches = FindAllModuleDeclarations(*tree);
+
+  // Nested module declarations are allowed, remove those
+  module_matches.erase(std::remove_if(module_matches.begin(),
+              module_matches.end(),
+              [](verible::TreeSearchMatch& m) {
+                  return m.context.IsInside(NodeEnum::kModuleDeclaration);
+              }),
+          module_matches.end());
```
Comment (fangism) [inline verilog/analysis/checkers/one_module_per_file_rule.cc]: Same as other file.

File: verilog/analysis/checkers/module_filename_rule.cc (hunk 3 above)
Comment (rkapuscik) [reply verilog/analysis/checkers/module_filename_rule.cc]: Done.

File: verilog/analysis/checkers/one_module_per_file_rule.cc (hunk 4 above)
Comment (rkapuscik) [reply verilog/analysis/checkers/one_module_per_file_rule.cc]: Done.

File: verilog/analysis/checkers/module_filename_rule.cc [hunk 5]
```diff
@@ -73,6 +74,20 @@ void ModuleFilenameRule::Lint(const TextStructureView& text_structure,
...
+  // If there are no modules in this source unit, suppress finding.
+  if (module_matches.empty()) return;
+
+  // Remove nested module declarations
+  std::vector<verible::TreeSearchMatch> module_cleaned;
+  module_cleaned.reserve(module_matches.size());
+  std::back_insert_iterator<std::vector<verible::TreeSearchMatch>> back_it(module_cleaned);
+  std::remove_copy_if(module_matches.begin(),
+                      module_matches.end(),
+                      back_it,
```
Comment (fangism) [inline verilog/analysis/checkers/module_filename_rule.cc]: FYI, can replace `back_it` with `std::back_inserter(module_cleaned)` (easier to write/read).

File: verilog/analysis/checkers/one_module_per_file_rule.cc [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+      return;
+  }
+
+  // Nested module declarations are allowed, remove those
+  std::vector<verible::TreeSearchMatch> module_cleaned;
+  module_cleaned.reserve(module_matches.size());
+  std::back_insert_iterator<std::vector<verible::TreeSearchMatch>> back_it(module_cleaned);
+  std::remove_copy_if(module_matches.begin(),
+                      module_matches.end(),
+                      back_it,
```
Comment (fangism) [inline verilog/analysis/checkers/one_module_per_file_rule.cc]: Same suggestion.
